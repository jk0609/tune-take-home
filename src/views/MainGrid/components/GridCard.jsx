import React, { useState } from 'react';
import PropTypes from 'prop-types';
import { startCase } from 'lodash';
import { Line } from 'react-chartjs-2';

import StyledGridCardWrapper from '../styledComponents/StyledGridCardWrapper';

const GridCard = ({
  avatar,
  name,
  occupation,
  revenue,
  impressions,
  conversions
}) => {
  const [imageError, setImageError] = useState(false);

  // generates a random background color for the blank avatar background
  const colors = ['red', 'yellow', 'blue', 'orange', 'green', 'purple', 'pink'];

  // Graphic configs
  const graphOptions = {
    legend: {
      display: false
    },
    scales: { xAxes: [{ display: false }], yAxes: [{ display: false }] },
    layout: {
      padding: 5
    }
  };

  let sortedDates = Object.keys(conversions.graphData).sort();
  const graphData = {
    labels: sortedDates,
    datasets: [
      {
        fill: false,
        backgroundColor: '#000',
        borderColor: '#000',
        pointRadius: 1,
        pointHitRadius: 10,
        data: sortedDates.map(day => conversions.graphData[day].toFixed(2))
      }
    ]
  };

  return (
    <StyledGridCardWrapper>
      <div className="card-header">
        <div className="card-avatar">
          {/* onError contains error handling for avatar images that are returning 403 responses */}
          {avatar && !imageError ? (
            <img
              src={avatar}
              alt="User avatar"
              onError={() => setImageError(true)}
            />
          ) : (
            <span
              style={{
                backgroundColor:
                  colors[Math.floor(Math.random() * colors.length)]
              }}
            >
              {name[0]}
            </span>
          )}
        </div>
        <div className="card-name">
          <h4>{name}</h4>
          <p>{startCase(occupation)}</p>
        </div>
      </div>
      <div className="card-body">
        <div className="card-graph">
          <Line data={graphData} options={graphOptions} />
          <span className="graph-label">{`Daily Revenue: ${sortedDates[0]} - ${
            sortedDates[sortedDates.length - 1]
          }`}</span>
        </div>
        <div className="card-metrics">
          <p className="impressions-count">{impressions}</p>
          <p className="impressions-label">impressions</p>
          <p className="conversions-count">{conversions.count}</p>
          <p className="conversions-label">conversions</p>
          <p className="revenue-count">
            {new Intl.NumberFormat('en-US', {
              style: 'currency',
              currency: 'USD'
            }).format(revenue)}
          </p>
        </div>
      </div>
    </StyledGridCardWrapper>
  );
};

GridCard.propTypes = {
  avatar: PropTypes.string,
  name: PropTypes.string,
  occupation: PropTypes.string,
  revenue: PropTypes.number,
  impressions: PropTypes.number,
  conversions: PropTypes.object
};

export default GridCard;
