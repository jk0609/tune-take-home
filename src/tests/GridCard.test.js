import React from 'react';
import Adapter from 'enzyme-adapter-react-16';
import { shallow, configure } from 'enzyme';
import toJson from 'enzyme-to-json';
import GridCard from '../views/MainGrid/components/GridCard';

configure({ adapter: new Adapter() });

jest.mock('react-chartjs-2', () => ({
  Line: () => null
}));

let wrapper;

let mockProps = {
  avatar: 'test-avatar-path',
  name: 'Test User',
  occupation: 'Test Occupation',
  revenue: 100,
  impressions: 20,
  conversions: {
    count: 10,
    graphData: {}
  }
};

beforeEach(() => {
  wrapper = shallow(<GridCard {...mockProps} />);
});

describe('GridCard snapshot', () => {
  it('should match snapshot', () => {
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
