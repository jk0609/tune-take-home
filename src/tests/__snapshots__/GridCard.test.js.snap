// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GridCard snapshot should match snapshot 1`] = `
<styled.div>
  <div
    className="card-header"
  >
    <div
      className="card-avatar"
    >
      <img
        alt="User avatar"
        onError={[Function]}
        src="test-avatar-path"
      />
    </div>
    <div
      className="card-name"
    >
      <h4>
        Test User
      </h4>
      <p>
        Test Occupation
      </p>
    </div>
  </div>
  <div
    className="card-body"
  >
    <div
      className="card-graph"
    >
      <Line
        data={
          Object {
            "datasets": Array [
              Object {
                "backgroundColor": "#000",
                "borderColor": "#000",
                "data": Array [],
                "fill": false,
                "pointHitRadius": 10,
                "pointRadius": 1,
              },
            ],
            "labels": Array [],
          }
        }
        options={
          Object {
            "layout": Object {
              "padding": 5,
            },
            "legend": Object {
              "display": false,
            },
            "scales": Object {
              "xAxes": Array [
                Object {
                  "display": false,
                },
              ],
              "yAxes": Array [
                Object {
                  "display": false,
                },
              ],
            },
          }
        }
      />
      <span
        className="graph-label"
      >
        Daily Revenue: undefined - undefined
      </span>
    </div>
    <div
      className="card-metrics"
    >
      <p
        className="impressions-count"
      >
        20
      </p>
      <p
        className="impressions-label"
      >
        impressions
      </p>
      <p
        className="conversions-count"
      >
        10
      </p>
      <p
        className="conversions-label"
      >
        conversions
      </p>
      <p
        className="revenue-count"
      >
        $100.00
      </p>
    </div>
  </div>
</styled.div>
`;
